File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/graph/P785_IsGraphBipartite2.java
package com.dsa.leetcode.labuladong.graph;

import java.util.LinkedList;
import java.util.Queue;

/**
 * @since 2024/6/30 09:09
 */
public class P785_IsGraphBipartite2 {
    boolean isBipartite = true;
    boolean[] colors;
    boolean[] visited;

    public boolean isBipartite(int[][] graph) {
        int n = graph.length;
        colors = new boolean[n];
        visited = new boolean[n];
        for (int i = 0; i < n; i++) {
            if (!isBipartite || !visited[i]) {
                bfs(graph, i);
            }
        }
        return isBipartite;
    }

    private void bfs(int[][] graph, int i) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(i);
        while (!queue.isEmpty()) {
            int from = queue.poll();
            visited[from] = true;
            for (int to : graph[from]) {
                if (!visited[to]) {
                    // change color
                    colors[to] = !colors[from];
                    queue.offer(to);
                } else {
                    // check "to" node's color
                    if (colors[to] == colors[from]) {
                        isBipartite = false;
                    }
                }
            }
        }
    }

    public static void main(String[] args) {
        int[][] graph = {{1, 2, 3}, {0, 2}, {0, 1, 3}, {0, 2}};
        P785_IsGraphBipartite2 p785 = new P785_IsGraphBipartite2();
        boolean bipartite = p785.isBipartite(graph);
        System.out.println(bipartite);
    }

}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/graph/P785_IsGraphBipartite2.py
from collections import deque
from typing import List


class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        num = len(graph)
        self.is_bipartite = True
        self.colors = [False] * num
        self.visited = [False] * num

        for i in range(num):
            if not self.is_bipartite:
                break
            if not self.visited[i]:
                self.bfs(graph, i)
        return self.is_bipartite

    def bfs(self, graph: List[List[int]], i: int):
        q = deque()
        q.append(i)
        while q:
            _from = q.popleft()
            self.visited[_from] = True
            for _to in graph[_from]:
                if not self.visited[_to]:
                    self.colors[_to] = not self.colors[_from]
                    q.append(_to)
                else:
                    if self.colors[_to] == self.colors[_from]:
                        self.is_bipartite = False
